### DISCUSSION TITLE
Add a chapter to handbook documentation about image data structures

### DISCUSSION THREAD

**Discussion by nopria** [python-pillow/Pillow#6828] [Q&A]
It would be very useful to know, for each image mode, how image data is structured. For some kind of images it is quite obvious, but for others, **especially if transparency is present**, it is not (or at least I was not able to find it).

For example, I searched in documentation an explanation to [this code](https://stackoverflow.com/a/58567453/694360) (in order to check if it's correct and to adapt it to my needs) but I could find only sparse and partial information.

Thanks in advance for your attention.

**Comment by radarhere** [comment]
I imagine you've seen https://pillow.readthedocs.io/en/stable/handbook/concepts.html. I would hope that explains that RGBA is made up of four channels ranging from 0 to 255, where the fourth channel is alpha.

I've created PR #6831 to add a "Transparency" section to the documentation - https://pillow--6831.org.readthedocs.build/en/6831/handbook/concepts.html#transparency

Let us know if that is sufficient, or there are other items that are unclear.

**Reply by nopria** [reply]
Thanks a lot, that's a great clarification, I just would explain briefly what's the effect of *premultiplied alpha* on image data.

My objective is to write a function able to remove transparency (keeping the image mode) and a function able to "optimize" image data structure by downgrading, only if possible, image mode (for example RGB->P->L->1, or RGBA->PA->LA), keeping image content intact. But I doubt that objective is beyond my skills, do you think those functionalities deserve opening a dedicated feature request?

**Reply by radarhere** [reply]
I've created PR #6838 for premultiplied alpha.

That feature doesn't sound to me like it would be helpful to many users, so I don't think it's an ideal feature request. However, here is some code that should cover most images.

```python
from PIL import Image

def isBlackAndWhite(im):
    for _, pixel in im.getcolors():
        if pixel not in (0, 255):
            return False
    return True

def downgrade(im):
    if im.mode in ("RGB", "RGBA"):
        channels = im.split()
        r, g, b = channels[:3]
        if r.tobytes() == g.tobytes() and g.tobytes() == b.tobytes():
            # R, G and B are the same. The RGB image is greyscale
            if im.mode == "RGB" and isBlackAndWhite(r):
                # If there is no transparency, and is black and white,
                # then it can be converted to 1
                return im.convert("1")
            return im.convert("L" if im.mode == "RGB" else "LA")
        rgb = im.convert("RGB")
        if rgb.getcolors() is not None:
            # There are 256 colors or less in the RGB image
            p = Image.new("P", im.size)
            for x in range(im.width):
                for y in range(im.height):
                    pixel = rgb.getpixel((x, y))
                    color = p.palette.getcolor(pixel)
                    p.putpixel((x, y), color)
            if im.mode == "RGBA":
                p.putalpha(channels[3])
                return p
            return p
    elif im.mode in ("P", "PA"):
        blackAndWhite = True
        for color in im.palette.colors:
            if color[0] != color[1] or color[1] != color[2]:
                break
            if color[0] not in (0, 255):
                blackAndWhite = False
        else:
            # P image is greyscale
            if im.mode == "PA":
                return im.convert("LA")
            else:
                return im.convert("1" if blackAndWhite else "L")
    elif im.mode == "L" and isBlackAndWhite(im):
        return im.convert("1")
    return im

assert downgrade(Image.new("RGB", (1, 1), (0, 0, 0))).mode == "1"
assert downgrade(Image.new("RGB", (1, 1), (1, 1, 1))).mode == "L"
assert downgrade(Image.new("RGB", (1, 1), (1, 1, 2))).mode == "P"
assert downgrade(Image.new("RGBA", (1, 1), (0, 0, 0, 0))).mode == "LA"
assert downgrade(Image.new("RGBA", (1, 1), (1, 1, 2, 0))).mode == "PA"
assert downgrade(Image.new("P", (1, 1))).mode == "1"
assert downgrade(Image.new("PA", (1, 1))).mode == "LA"
assert downgrade(Image.new("L", (1, 1), 0)).mode == "1"
```

**Reply by nopria** [reply]
Thanks a lot, this function seems great.

I have one curiosity about the conversion from `RGB` with less than 256 colors to `P`. I see that the image is recreated from scratch, I wonder if the same result could be obtained with
```python
Image.convert(mode='P', palette=Image.Palette.ADAPTIVE)
```
I checked with a couple of images and it seems to work, i.e. the colors of resulting `P` image are exactly the same as the `RGB` original (whereas without the `palette=Image.Palette.ADAPTIVE` parameter the colors change because of quantization).

**Reply by radarhere** [reply]
I've been simultaneously answering #6832, where a user is discussing a change in pixels when converting from RGBA to P. But that is from RGBA to P, not RGB to P.

Testing, I also find that `im.convert("P", palette=Image.Palette.ADAPTIVE)` doesn't cause any loss for less than 256 colors. So sure, you can adjust the code to use that for RGB images.

**Reply by radarhere** [reply]
PR #6838 has been merged. Is this question answered?

**Reply by nopria** [reply]
Sure, answered beyond my most optimistic hope!

And I still think that the `downgrade` function above would be a useful core command...

**Reply by radarhere** [reply]
Why are you interested in this functionality? At first I thought maybe you were trying to minimise the output of `tobytes()` for storage purposes, but if you were trying to optimise storage, you'd be better off using an image format like PNG.

**Reply by nopria** [reply]
Since I have to produce PDF 1.3 (for printing purposes) with all kind of different images (photos, drawings, etc.) and the images must be not less than 300 DPI of real printing size, but also not much more than 300 DPI (to keep the PDF -- usually hundreds of pages -- as small as possible), my workflow force me to:
- downsample (seldomly upscale) each image at the number of pixels necessary to obtain just not less than 300 DPI in printing
- convert to sRGB colorspace and remove transparencies (PDF 1.3 does not support transparent images)
- [identify almost greyscale images](https://stackoverflow.com/q/74888295/694360) (usually color scan of greyscale originals) and force them to greyscale, which they benefit from
- quantize to about 32 colors images like drawings without graduated shadings, which quality often benefits from strong quantization
- convert all image formats to either JPG (photos) or PNG (any image with less than 256 colors)

In such a mess I just find very useful to have a way to convert image mode **losslessly** (i.e. without altering image content, optimizing the data structure of image if downgrading) just before saving it to JPG or PNG.

This `downgrade` function might be added to Pillow `convert` command easily, just adding a `force` parameter which would default to `True`, and users like me who want to convert to a specific image mode without altering image content could set `force=False`.

**Reply by radarhere** [reply]
Not all image formats support all modes when saving, so I'm concerned that converting to a lower mode might actually introduce an error in some circumstances. Perhaps if #2663 is solved, then that would provide a good basis for some sort of general `optimize` argument when saving.

> This `downgrade` function might be added to Pillow `convert` command easily, just adding a `force` parameter which would default to `True`, and users like me who want to convert to a specific image mode without altering image content could set `force=False`.

Although here you seem to not be talking about a method to downgrade the mode as much as possible, but instead altering `convert()` so that if told to only be lossless, and the image isn't simple enough for that... it just returns the image unchanged?

**Reply by nopria** [reply]
> Although here you seem to not be talking about a method to downgrade the mode as much as possible, but instead altering `convert()` so that if told to only be lossless, and the image isn't simple enough for that... it just returns the image unchanged?

Yes, I admit my explanation wasn't clear, because in my mind your `downgrade` function is the *above expectations* objective of a lossless mode `convert` command, i.e. a conversion affecting only image data structure leaving intact image content.
